invoices/invoice_expiry_watcher.go

```diff
@@ -0,0 +1,193 @@
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/queue"
+	"github.com/lightningnetwork/lnd/zpay32"
+)
+
+// invoiceExpiry holds and invoice's payment hash and its expiry. This
+// is used to order invoices by their expiry for cancellation.
+type invoiceExpiry struct {
+	PaymentHash lntypes.Hash
+	Expiry      time.Time
+}
+
+// Less implements PriorityQueueItem.Less such that the top item in the
+// priorty queue will be the one with that expires next.
+func (e invoiceExpiry) Less(other queue.PriorityQueueItem) bool {
+	return e.Expiry.Before(other.(*invoiceExpiry).Expiry)
+}
+
+// InvoiceExpiryWatcher handles automatic invoice cancellation of expried
+// invoices.
+// Upon start InvoiceExpiryWatcher will subscribe to all invoice events and
+// will add all active (not yet settled and not yet cancelled) invoices to its
+// watcing queue.
+// When a new invoice is added to the InvoiceRegistry, InvoiceExpiryWatcher
+// will add that to the watching queue as well.
+// If any of the watched invoices expire, they'll be removed from the watching
+// queue and will be cancelled through InvoiceRegistry.CancelInvoice.
+type InvoiceExpiryWatcher struct {
+	sync.Mutex
+	started bool
+	wg      sync.WaitGroup
+
+	// clock is the clock implementation that InvoiceExpiryWatcher uses.
+	// It is useful for testing.
+	clock clock.Clock
+
+	// quit signals InvoiceExpiryWatcher to stop.
+	quit chan interface{}
+
+	// cancelInvoice is a template method that cancels an expired invoice.
+	cancelInvoice func(lntypes.Hash) error
+
+	// expiryQueue holds invoiceExpiry items and is used to find the next
+	// invoice to expire.
+	expiryQueue queue.PriorityQueue
+
+	// newInvoices channel is used to wake up the main loop when a new invoices
+	// is added.
+	newInvoices chan *invoiceExpiry
+}
+
+// NewInvoiceExpiryWatcher creates a new InvoiceExpiryWatcher instance.
+func NewInvoiceExpiryWatcher(clock clock.Clock) *InvoiceExpiryWatcher {
+	return &InvoiceExpiryWatcher{
+		clock:       clock,
+		quit:        make(chan interface{}),
+		newInvoices: make(chan *invoiceExpiry),
+	}
+}
+
+// PrefetchInvoices fetches all active invoices and their corresponding payment
+// hashes from ChannelDB and adds them to the watcher. This is useful to
+// prepopulate the watcher with past, but active invoices upon start.
+func (ew *InvoiceExpiryWatcher) PrefetchInvoices(cdb *channeldb.DB) {
+	pendingOnly := true
+	invoices, err := cdb.FetchAllInvoices(pendingOnly)
+	if err != nil {
+		log.Errorf("Error fetching invoices from the database: %v", err)
+	} else {
+		for k := 0; k < len(invoices); k++ {
+			var paymentHash lntypes.Hash
+			copy(paymentHash[:lntypes.HashSize], invoices[k].PaymentHash)
+			ew.AddInvoice(paymentHash, &invoices[k])
+		}
+	}
+}
+
+// Start starts the the subscription handler and the main loop. Start() can
+// only be called once.
+func (ew *InvoiceExpiryWatcher) Start(
+	cancelInvoice func(lntypes.Hash) error) error {
```

Comment:
Any particular reason we pass the cancel closure here instead of in the main constructor? 

Following from this and the above comment, we can introduce a pretty slim interface here to handle fetching+cancelling the set of invoices.

Reply:
As `InvoiceExpiryWatcher` is a dependency of `InvoiceRegistry` it is passed during construction and needs to exist before the registry.